drivers/mipi_dbi/mipi_dbi_rpi_pico_pio.c

```diff
@@ -0,0 +1,535 @@
+}
+
+static int mipi_dbi_pico_pio_write_helper(const struct device *dev,
+					  const struct mipi_dbi_config *dbi_config,
+					  bool cmd_present, uint8_t cmd, const uint8_t *data_buf,
+					  size_t len)
+{
+	int ret = 0;
+
+	const struct mipi_dbi_pico_pio_config *config = dev->config;
+	struct mipi_dbi_pico_pio_data *data = dev->data;
+
+	ret = k_mutex_lock(&data->lock, K_FOREVER);
+	if (ret < 0) {
+		return ret;
+	}
+
+	switch (dbi_config->mode) {
+	case MIPI_DBI_MODE_8080_BUS_8_BIT:
+	case MIPI_DBI_MODE_8080_BUS_9_BIT:
+	case MIPI_DBI_MODE_8080_BUS_16_BIT:
+		gpio_pin_set_dt(&config->cs, 1);
+		if (cmd_present) {
+			pio_set_sm_mask_enabled(data->pio, data->sm_mask, false);
+			gpio_pin_set_dt(&config->cmd_data, 0);
+			for (int i = 0; i < data->split_count; ++i) {
+				pio_sm_put_blocking(data->pio, data->split[i].sm.sm, (uint32_t)cmd);
+			}
+			pio_enable_sm_mask_in_sync(data->pio, data->sm_mask);
+			mipi_dbi_pico_pio_wait_end_of_transmission(data);
+			gpio_pin_set_dt(&config->cmd_data, 1);
+		}
+		if (len > 0) {
+			k_msgq_purge(config->msq_dma);
+
+			pio_set_sm_mask_enabled(data->pio, data->sm_mask, false);
+
+			uint32_t dma_channel_mask = 0;
+
+			for (int i = 0; i < data->split_count; ++i) {
+				WRITE_BIT(dma_channel_mask, data->split[i].dma.channel, true);
+				data->split[i].dma.head_block.block_size = len;
+				data->split[i].dma.head_block.source_address =
+					(uint32_t)&data_buf[0];
+				dma_config(config->dev_dma, data->split[i].dma.channel,
+					   &data->split[i].dma.config);
+				dma_start(config->dev_dma, data->split[i].dma.channel);
+			}
+
+			pio_enable_sm_mask_in_sync(data->pio, data->sm_mask);
+
+			/* wait for end of dma(s) */
+			int channel;
+
+			while (dma_channel_mask != 0) {
+				k_msgq_get(config->msq_dma, &channel, K_FOREVER);
+				WRITE_BIT(dma_channel_mask, channel, false);
+			}
+
+			mipi_dbi_pico_pio_wait_end_of_transmission(data);
+		}
+		gpio_pin_set_dt(&config->cs, 0);
+		break;
+
+	default:
+		LOG_ERR("MIPI DBI mode %u is not supported.", dbi_config->mode);
+		ret = -ENOTSUP;
+	}
+
+	k_mutex_unlock(&data->lock);
+	return ret;
+}
+
+static int mipi_dbi_pico_pio_command_write(const struct device *dev,
+					   const struct mipi_dbi_config *dbi_config, uint8_t cmd,
+					   const uint8_t *data_buf, size_t len)
+{
+	return mipi_dbi_pico_pio_write_helper(dev, dbi_config, true, cmd, data_buf, len);
+}
+
+static int mipi_dbi_pico_pio_write_display(const struct device *dev,
+					   const struct mipi_dbi_config *dbi_config,
+					   const uint8_t *framebuf,
+					   struct display_buffer_descriptor *desc,
+					   enum display_pixel_format pixfmt)
+{
+	ARG_UNUSED(pixfmt);
+
+	return mipi_dbi_pico_pio_write_helper(dev, dbi_config, false, 0x0, framebuf,
+					      desc->buf_size);
+}
+
+static int mipi_dbi_pico_pio_reset(const struct device *dev, k_timeout_t delay)
+{
+	const struct mipi_dbi_pico_pio_config *config = dev->config;
+	int ret;
+
+	LOG_DBG("Performing hw reset.");
```

Comment:
Maybe this message could say more about which HW is reset, e.g.:

```suggestion
	LOG_DBG("Resetting the display.");
```